testing-guidelines/ci.md

```diff
@@ -0,0 +1,431 @@
+
+By default all checks run in Linux, but we run certain Pyhon checks in windows. Sadly, Windows action runners are very slow and the windows tests are mostly there to handle a few edge cases. In order to ensure that we have good test coverage but don't slow down contributors with long CI runs, we only run the Windows check in the release pull request. If a pull request is specifically targetting a Windows issue, then maintainers can add the `windows` label which will make the Windows test run in that pull request as well.
+
+We only run our Python tests for a single Python version, the oldest we support (3.8), for practical reasons.
+
+##### Checking out branches
+
+As we use `workflow_run` events for most jobs, we don't have easy access via the workflow context of the branch we actually care about. The `find-pr` action returns information not just about the pull request but also about the branch associated with it, importantly this includes both the HEAD SHA of that branch and the merge SHA. The merge SHA is the merged result of the branch and it's target and this is almost always the branch that we checkout, as it gives us confidence that the feature works when merged into `main` and not only in the PR branch.
+
+The one exception to this is the visual tests. We use an external services to handle visual testing and storybook deployments ([chromatic](https://www.chromatic.com)) and they prefer the head commit of the branch to be used, so in this case we check out the HEAD SHA instead of the merge SHA.
+
+##### Deploying previews
+
+We deploy up to 3 ephemeral 'environments' and build a new wheel for gradio when a pull request is created:
+
+- `storybook` - previews any frontend components that have a `.stories.svelte` file.
+- `website` - previews the website - the `main` version will show any docs or guide changes present in the pull request.
+- `spaces` - previews a series of gradio apps on Huggin Face Spaces as they will appear and behave when the pull requests is merged and released.
+
+The `storybook` and `website` previews are relatively straightforward as they use tooling provided by the service providers that we use in order to deploy. We just make sure each run has the necessary assets and components built.
+
+The `spaces` preview is a little more involved as it is a custom process and requires us to build a custom gradio wheel from that pull request.
+
+The process is relatively straightforward, and follows [the steps mentioned above](#spaces) but there a few details to be aware of.
+
+- We use [a custom script](https://github.com/gradio-app/gradio/blob/main/scripts/copy_demos.py) to pull in a select number of spaces and build them into a single FastAPI application. We serve each demo on its own subpath. This is the demo app that gets deployed to spaces.
+- We build a new wheel from the pull requests source code and upload it to s3, we then add the url for this wheel to the requirements.txt of the space we are deploying.
```

Comment:
FWIW, https://github.com/aws-actions/configure-aws-credentials supports GitHub's OIDC too these days, so no need for secret tokens...